typing/typedecl_variance.ml

```diff
@@ -329,16 +293,17 @@ let compute_variance_decl env ~check decl (required, _ as rloc) =
           mn @ List.map (fun c -> c.Types.cd_args, c.Types.cd_res) tll in
         match List.map (compute_variance_gadt env ~check rloc decl) tll with
         | vari :: rem ->
-            let varl = List.fold_left (List.map2 Variance.union) vari rem in
-            List.map
-              Variance.(fun v -> if mem Pos v && mem Neg v then full else v)
-              varl
+            List.fold_left (List.map2 Variance.union) vari rem
         | _ -> assert false
       end
   | Type_record (ftl, _) ->
       compute_variance_type env ~check rloc decl
         (mn @ List.map (fun {Types.ld_mutable; ld_type} ->
              (ld_mutable = Mutable, ld_type)) ftl)
+  in
+  if mn = [] || decl.type_kind <> Type_abstract then
+    List.map Variance.strengthen vari
+  else vari
```

Comment:
Re-reading this code that I had forgotten, I have a hard time reconstructing the meaning of the three lines:

```ocaml
  if mn = [] || decl.type_kind <> Type_abstract then
    List.map Variance.strengthen vari
  else vari
```

On March 1st I wrote

> here you are generalizing again covariant variables in non-strictly-positive but injective positions).

I believe that it is correct, but I have no clue how I guessed that from the code.

In any case, could you maybe add a comment to explain what we are doing here?

Also: where was this logic placed before? Do I correctly understand that this new code corresponds to the following lines of the old version that are now removed?

```ocaml
let compute_variance_type ... =
  ... 
  List.map2
    (...
      if decl.type_kind = Type_abstract && tr = Public then v else
      set May_weak (mem May_neg v) v)
  params required
```

Reply:
Concerning `Variance.strengthen`, it is used when the type cannot be expanded `mn = []` (or expanding it would not add information, since it is a datatype). The idea is that in that case there is no way to replace a positive occurrence by a succession of two negative occurrences, so that the above counter-example does not apply.

I think you are correct that the original code did that strengthening earlier.
Since this a property of the whole type definition, it seems more natural to do it in one pass here.
I could add a comment on the meaning of this pass.